hook_macro/src/lib.rs

```diff
@@ -0,0 +1,72 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, AttributeArgs};
+use quote::ToTokens;
+
+
+#[proc_macro_attribute]
+/// This macro is used to insert a hook at the start of a function.
+pub fn cvt_hook_start(attr: TokenStream, input: TokenStream) -> TokenStream {
+    
+    // parse the attribute argument
+    let attr = parse_macro_input!(attr as AttributeArgs);
+    if attr.len() != 1 {
+        panic!("expected 1 argument");
+    }
+
+    let arg = &attr[0]; 
+
+    // parse the input tokens and make sure it is a function
+    let mut item: syn::Item = syn::parse(input).unwrap();
```

Comment:
why not parse as Item::Fn?

Reply:
parsing it as `Item` and then checking if it is an `Item::Function` allows to give a better error message in case the hook is applied to _non-function_ item (e.g. Struct)

Reply:
Is that a standard practice? I haven't seen this before, hence the comment. But I'm not an expert either ;)